src/session.js

```diff
@@ -0,0 +1,134 @@
+        startTime: startTimeAsNewDate_getTime
+    }
+    Future work can add more fields though note that the fact that the value doesn't change 
+    once created makes this very robust when used in multiple tabs/windows - tabs don't compete/
+    race to do anything but set the max-age.
+*/
+
+
+const MaxSessionAgeMillis = 4 * 60 * 60 * 1000;
+const InactivityTimeoutSeconds = 15 * 60;
+const PeriodicCheckSeconds = 60;
+export const COOKIE_NAME = '_splunk_rum_sid';
+
+let rumSessionId = generateId(64); // will be overwritten in init() with scriptInstance; only here as extreme fallback
+let recentActivity = false;
+function markActivity() {
+  recentActivity = true;
+}
+  
+function pastMaxAge(startTime) {
+  return startTime > new Date().getTime() || new Date().getTime() > startTime + MaxSessionAgeMillis;
+}
+  
+function parseCookieToSessionState() {
+  const rawValue = findCookieValue(COOKIE_NAME);
+  if (!rawValue) {
+    return undefined;
+  }
+  const decoded = decodeURIComponent(rawValue);
+  if (!decoded) {
+    return undefined;
+  }
+  let ss = undefined;
+  try {
+    ss = JSON.parse(decoded);
+  } catch (error) {
+    return undefined;
+  }
+  // should exist and be an object
+  if (!ss || typeof ss !== 'object') {
+    return undefined;
+  }
+  // id validity
+  if (!ss.id || typeof ss.id !== 'string' || !ss.id.length || ss.id.length !== 32) {
+    return undefined;
+  }
+  // startTime validity
+  if (!ss.startTime || typeof ss.startTime !== 'number' || pastMaxAge(ss.startTime)) {
+    return undefined;
+  }
+  return ss;
+}
+
+function newSessionState() {
+  return {
+    id: generateId(128),
+    startTime: new Date().getTime()
+  };
+}
+
+function renewCookieTimeout(sessionState) {
+  if (pastMaxAge(sessionState.startTime)) { // safety valve
+    return;
+  }
+  const cookieValue = encodeURIComponent(JSON.stringify(sessionState));
+
+  let cookie = COOKIE_NAME + '=' + cookieValue + '; path=/; max-age=' + InactivityTimeoutSeconds ;
+
+  if (isIframe()) {
+    cookie += ';SameSite=None; Secure';
+  } else {
+    cookie += ';SameSite=Strict';
+  }
+  document.cookie = cookie;
+}
+
+// This is called periodically and has two purposes:
+// 1) Check if the cookie has been expired by the browser; if so, create a new one
+// 2) If activity has occured since the last periodic invocation, renew the cookie timeout
+// (Only exported for testing purposes.)
+export function updateSessionStatus() {
+  let sessionState = parseCookieToSessionState();
+  if (!sessionState) {
+    sessionState = newSessionState();
+    recentActivity = true;  // force write of new cookie
+  }
+  rumSessionId = sessionState.id;
+  if (recentActivity) {
+    renewCookieTimeout(sessionState);
+  }
+  recentActivity = false;
+}
+
+export function initSessionTracking(instanceId) {
+  rumSessionId = instanceId;
+  recentActivity = true; // document loaded implies activity
+  updateSessionStatus();
+  setInterval(updateSessionStatus, PeriodicCheckSeconds * 1000);
```

Comment:
Looks like a ducktape for _.throttle (<https://lodash.com/docs/4.17.15#throttle>). For our case I would add `{leading: true, trailing: true}`, which should mirror activity callbacks more closely.

Reply:
I'd just as soon not pull in another library unless the complexity of this thing grows substantially bigger.  Also, it's important for the behavior of this that the first `updateSessionStatus` happen before `init` returns, rather than on next tick...